sunpy/data/manager/cache.py

```diff
@@ -0,0 +1,93 @@
+from urllib.request import urlopen
+
+from sunpy.util.util import hash_file, replacement_filename
+from sunpy.util.net import get_filename
+
+
+class Cache:
+    """
+    Cache handles caching.
+    """
+
+    def __init__(self, downloader, storage, cache_dir):
+        self._downloader = downloader
+        self._storage = storage
+        self._cache_dir = cache_dir
+        if not self._cache_dir.endswith('/'):
+            self._cache_dir += '/'
+
+    def download(self, urls, redownload=False):
+        """
+        Downloads the files from the urls.
+
+        Parameters
+        ----------
+        urls: `list`
+            A list of urls.
+        redownload: `bool`
+            Whether to skip cache and redownload.
+        """
+        # TODO: Expiry time
+        # XXX: Expiry time cache level or download level?
+        if not redownload:
+            details = self._get_by_url(urls[0])
+            if details:
+                return details['file_path']
+
+        file_path, file_hash, url = self._download_and_hash(urls)
+
+        if not redownload:
+            self._storage.store({
+                'file_hash': file_hash,
+                'file_path': file_path,
+                'url': url,
+            })
```

Comment:
This `if` block only happens if `_get_by_url()` on l33 doesn't return `details`... but it's been redownloaded on l37, right?

Reply:
Yes. I was hesitant on splitting if logic like this, but I think this makes it more readable.

Reply:
I need to get a better gist to know what's going on... but I would think the `_storage.store` should be updated in any case, right? Why won't return the details above? because the files were there but not hashed?

Reply:
> the _storage.store should be updated in any case, right? 

No. Not when the cache has the file downloaded. 

If the user wants to redownload, the file should not go in cache because there is a chance of hash collision.

Does that make sense? That's what I had thought when I wrote it I think.

Reply:
A flow for more insight and future reference:

1. If `redownload`: Don't check cache, don't put in cache. Download and return file path
2. If not `redownload`: Check cache, 
	1. If present in cache: return path
	2. If not download, store in cache and return path

Reply:
Ahh! That makes sense now. I hadn't understood that if redownloaded it was not put in the cache. I don't understand why one wouldn't want to put it there, but knowing these steps helps me to understand the code.
In any case, to keep this as future reference, then it should be as comments or docstrings. Nobody is going to look in here in 10 years time. :stuck_out_tongue_closed_eyes: